### PR TITLE
hammer: osd: fix fiemap issue in xfs when #extents > 1364

### PR DESCRIPTION
**Pull request by mslovy**
http://tracker.ceph.com/issues/17878

backport: for issue #17610

### CODE REVIEW

**Comment by smithfarm** [comment]
@mslovy Please rebase to get rid of the "This branch is out-of-date" warning.

**Comment by smithfarm** [comment]
@mslovy While you're rebasing, please re-do the cherry-pick of 1a1c126d80b427d5230347fbc71a4edea5c0b6c8 using `git cherry-pick -x` -- the "(cherry picked from . . .)" line is needed in the commit message.

**Comment by mslovy** [comment]
redo cherry-pick and solve conflict,  please review  again

**Comment by smithfarm** [comment]
@mslovy Thanks for the backport! I will test it.

**Comment by smithfarm** [comment]
@liewegas This PR passed a rados run at http://tracker.ceph.com/issues/17151#note-23

Do you think it can be merged?

**Comment by tchaikov** [comment]
might want to cherry-pick https://github.com/ceph/ceph/pull/12148/commits/c3748fa7737b6c43de2dae5d957fcff01322515c also once #12148 is merged.

**Comment by liewegas** [comment]
let's make sure this is merged along with the use-after-free fix.  otherwise lgtm!

**Comment by tchaikov** [comment]
#12148 is merged, and the commit to be cherry-picked in that PR is c3748fa7737b6c43de2dae5d957fcff01322515c

**Comment by smithfarm** [comment]
@mslovy Can you cherry-pick c3748fa into this PR?

**Comment by mslovy** [comment]
okey

File: src/os/FileStore.cc [hunk 1]
```diff
@@ -2913,7 +2913,7 @@ int FileStore::fiemap(coll_t cid, const ghobject_t& oid,
   } else {
     uint64_t i;
     struct fiemap_extent *extent = NULL;
-    struct fiemap_extent *last = NULL;
+    struct fiemap_extent *last = nullptr;
```
Comment (smithfarm) [inline src/os/FileStore.cc]: please change "nullptr" to 0 (zero) because nullptr is a C++11ism

Comment (smithfarm) [inline src/os/FileStore.cc]: or just leave the NULL

**Comment by smithfarm** [comment]
@mslovy hammer uses C++98 so nullptr will not work, please change it to "NULL" and re-push.

**Comment by mslovy** [comment]
got it

**Comment by smithfarm** [comment]
@athanatos This just passed a rados run at http://tracker.ceph.com/issues/17151#note-38 and it includes the use-after-free fix @liewegas mentioned at https://github.com/ceph/ceph/pull/11615#issuecomment-263280908

Do you think it's OK to merge?

**Pull request merged**